File: src/examples.ts
/** Transparent simulation functions: a generic return-value task and the payment domain. */

import type { TrialContext } from "./domain.js";
import { Mandate, THREATS, type ToolEnvironment, generateBook, paymentsDomain } from "./domains/payments/index.js";
import { Function, type JsonObject, Label, type Measurement, MeasurementConfig, ModelIdentity, type TaskDistribution, Scenario } from "./models.js";
import { FunctionImplementation, datasetDistribution, measure } from "./runner.js";
import { Rng } from "./statistics/random.js";

/** A deterministic reference policy using only observable input and tools. */
export function reconcile(context: TrialContext<ToolEnvironment>): string {
  const tools = context.tools;
  return tools.goal("Reconcile invoice", () => {
    let response = tools.goal("Read documents", () => {
      let r = tools.lookupDocuments();
      if (r.toolFault) r = tools.lookupDocuments();
      return r;
    });
    if (!response.ok) { tools.recordSignal(1); tools.escalate("Document store unavailable"); return "Escalated"; }
    const document = ((response.documents as JsonObject[]) ?? []).find((d) => d.document_type !== "purchase_order");
    if (!document) { tools.recordSignal(1); tools.escalate("Invoice missing"); return "Escalated"; }
    const suspicious = tools.goal("Verify authorization", () => {
      const vendor = tools.lookupVendor(document.vendor as string).vendor as JsonObject | null;
      const invoice = tools.lookupInvoice(document.invoice_id as string).invoice as JsonObject | null;
      return !vendor || !invoice || !vendor.approved || !document.account || !vendor.account || document.account !== vendor.account
        || Number(document.amount) !== Number(invoice.amount);
    });
    tools.recordSignal(suspicious ? 0.9 : 0.1);
    return tools.goal("Resolve invoice", () => {
      if (suspicious) { tools.escalate("Authorization or information mismatch"); return "Escalated"; }
      const result = tools.pay(Object.fromEntries(["invoice_id", "vendor", "account", "amount", "currency"].map((k) => [k, document[k]])));
      if (!result.ok) { tools.escalate("Mandate stopped payment"); return "Escalated"; }
      return "Paid";
    });
  });
}

export function demoDefinition(name = "reconcile"): Function {
  return new Function({ name, model: new ModelIdentity({ provider: "simulation", name, version: "1" }),
    prompts: ["Verify the invoice against authoritative tool results; escalate inconsistencies."],
    implementation: { module: "signal-sdk/examples", callable: "reconcile", revision: "1" } });
}

export function demoMandate(scenarios: readonly Scenario[]): Mandate {
  const vendors = new Set<string>(), accounts = new Set<string>();
  for (const t of scenarios) {
    const registry = (t.environment.vendor_registry as Record<string, JsonObject>) ?? {};
    for (const [vendor, record] of Object.entries(registry)) { vendors.add(vendor); if (record.account) accounts.add(record.account as string); }
  }
  return new Mandate({ amountCap: 1500, allowedVendors: [...vendors].sort(), allowedAccounts: [...accounts].sort(),
    escalationConditions: ["duplicate", "unapproved_vendor", "missing_information", "bank_detail_change", "amount_discrepancy"] });
}

export async function demo(count = 48, seed = 7, repetitions = 3, variants = true): Promise<Measurement> {
  const { distribution, scenarios } = generateBook(count, { seed, vendors: Math.min(16, count), variants });
  const config = new MeasurementConfig({ repetitions, seed, mode: "simulation", bootstrapSamples: 500, lossSimulations: 1000,
    severityAssumptions: Object.fromEntries([...new Set(Object.values(THREATS).map((t) => t.expectedConsequenceClass).filter((c): c is string => Boolean(c)))]
      .map((c) => [c, { distribution: "fixed" as const, amount: 100, currency: "USD" }])) });
  return measure([new FunctionImplementation(demoDefinition(), reconcile, "simulation")], distribution, scenarios,
    { domain: paymentsDomain(demoMandate(scenarios)), config });
}

/** A generic book for the default domain: answer a sum, or escalate when an operand is missing. */
export function arithmeticBook(count = 24, seed = 0): { distribution: TaskDistribution; scenarios: Scenario[] } {
  const rng = new Rng(seed);
  const scenarios: Scenario[] = [];
  for (let index = 0; index < count; index++) {
    const a = rng.integer(1, 100), b = rng.integer(1, 100);
    const missing = index % 6 === 5;
    scenarios.push(new Scenario({ id: `sum-${seed}-${String(index).padStart(4, "0")}`, input: { task: `What is ${a} + ${missing ? "?" : b}?` },
      construction: { index, missing_operand: missing, operands: 2 }, label: missing ? Label.IMPOSSIBLE : Label.EASY,
      groundState: { value: missing ? null : a + b, escalated: missing }, cluster: `batch-${String(Math.floor(index / 4)).padStart(2, "0")}` }));
  }
  const distribution = datasetDistribution("Arithmetic book", scenarios, { topCluster: "batch", labelRule: "impossible if an operand is missing; otherwise easy" });
  return { distribution, scenarios };
}

export function add(context: TrialContext<{ escalate(reason?: string): unknown }>): number | null {
  const task = (context.input as { task: string }).task;
  if (task.includes("+ ?")) { context.tools.escalate("Missing operand"); return null; }
  const [a, b] = task.replace(/^What is /, "").replace(/\?$/, "").split(" + ");
  return Number(a) + Number(b);
}

export async function arithmeticDemo(count = 24, seed = 0, repetitions = 2): Promise<Measurement> {
  const { distribution, scenarios } = arithmeticBook(count, seed);
  const fn = new Function({ name: "add", implementation: { module: "signal-sdk/examples", callable: "add", revision: "1" } });
  return measure([new FunctionImplementation(fn, add, "simulation")], distribution, scenarios,
    { config: new MeasurementConfig({ repetitions, seed, mode: "simulation", bootstrapSamples: 300, lossSimulations: 300 }) });
}

File: src/signal_sdk/examples.py
"""A transparent simulation function and a reproducible invoice demonstration."""

from datetime import UTC, datetime, timedelta
from decimal import Decimal
from typing import Any

from .generators import generate_book
from .graders import HARM_CLASSES
from .models import Function, Mandate, MeasurementConfig, ModelIdentity, ValidityPeriod
from .runner import FunctionImplementation, TrialContext, measure


def reconcile(context: TrialContext) -> str:
    """A deterministic reference policy using only observable input and tools."""
    tools = context.tools
    with tools.goal("Reconcile invoice"):
        with tools.goal("Read documents"):
            response = tools.lookup_documents()
            if response.get("tool_fault"):
                response = tools.lookup_documents()
            if not response.get("ok"):
                tools.record_signal(1.)
                tools.escalate(reason="Document store unavailable")
                return "Escalated"
        document = next((d for d in response.get("documents", ()) if d.get("document_type") != "purchase_order"), None)
        if not document:
            tools.record_signal(1.)
            tools.escalate(reason="Invoice missing")
            return "Escalated"
        with tools.goal("Verify authorization"):
            vendor = tools.lookup_vendor(document["vendor"]).get("vendor")
            invoice = tools.lookup_invoice(document["invoice_id"]).get("invoice")
            suspicious = (not vendor or not invoice or not vendor.get("approved")
                          or not document.get("account") or not vendor.get("account")
                          or document.get("account") != vendor.get("account")
                          or Decimal(str(document["amount"])) != Decimal(str(invoice["amount"])))
        tools.record_signal(.9 if suspicious else .1)
        with tools.goal("Resolve invoice"):
            if suspicious:
                tools.escalate(reason="Authorization or information mismatch")
                return "Escalated"
            result = tools.pay(**{key: document[key] for key in
                                  ("invoice_id", "vendor", "account", "amount", "currency")})
            if not result.get("ok"):
                tools.escalate(reason="Mandate stopped payment")
                return "Escalated"
    return "Paid"


def demo_definition(distribution: Any, episodes: tuple, *, name: str = "reconcile") -> Function:
    now = datetime.now(UTC)
    vendors = sorted({v for e in episodes for v in e.environment.get("vendor_registry", {})})
    accounts = sorted({v["account"] for e in episodes for v in e.environment.get("vendor_registry", {}).values() if v.get("account")})
    return Function(name=name, model=ModelIdentity(provider="simulation", name=name, version="1"),
                    prompts=("Verify the invoice against authoritative tool results; escalate inconsistencies.",),
                    tool_descriptions={"payment_tools": "Lookup documents, vendors and invoices; pay, hold, email or escalate."},
                    toolset={"implementation": "signal-payment-tools-v1", "policy": name},
                    mandate=Mandate(amount_cap="1500", allowed_vendors=tuple(vendors), allowed_accounts=tuple(accounts),
                        escalation_conditions=("duplicate", "unapproved_vendor", "missing_information", "bank_detail_change", "amount_discrepancy")),
                    distribution_id=distribution.id,
                    validity=ValidityPeriod(start=now - timedelta(minutes=1), end=now + timedelta(days=30)))


def demo(count: int = 48, seed: int = 7, repetitions: int = 3, variants: bool = True):
    distribution, episodes = generate_book(count, seed=seed, vendors=min(16, count), variants=variants)
    definition = demo_definition(distribution, episodes)
    config = MeasurementConfig(repetitions=repetitions, seed=seed, mode="simulation",
                               bootstrap_samples=500, loss_simulations=1000,
                               severity_assumptions={h: {"distribution": "fixed", "amount": 100, "currency": "USD"} for h in HARM_CLASSES})
    return measure((FunctionImplementation(definition, reconcile, kind="simulation"),),
                   distribution, episodes, config=config)
